src/core/protocol-authorization.ts

```diff
@@ -174,36 +169,40 @@ export class ProtocolAuthorization {
     inboundMessage: RecordsReadMessage | RecordsWriteMessage,
     protocolDefinition: ProtocolDefinition,
     ancestorMessageChain: RecordsWriteMessage[],
-    recordSchemaToLabelMap: Map<string, string>
   ): ProtocolRuleSet {
-    // make a copy of the ancestor messages and include the inbound write in the chain
-    const messageChain = [...ancestorMessageChain];
+    let protocolPath: string;
     if (inboundMessage.descriptor.method === DwnMethodName.Write) {
-      messageChain.push(inboundMessage as RecordsWriteMessage);
+      protocolPath = (inboundMessage as RecordsWriteMessage).descriptor.protocolPath!;
+    } else {
+      protocolPath = ancestorMessageChain[ancestorMessageChain.length-1].descriptor.protocolPath!;
     }
+    const protocolPathArray = protocolPath.split('/');
 
-    // walk down the ancestor message chain from the root ancestor record and match against the corresponding rule set at each level
-    // to make sure the chain structure is allowed
-    let allowedRecordsAtCurrentLevel: { [key: string]: ProtocolRuleSet} | undefined = protocolDefinition.records;
-    let currentMessageIndex = 0;
-    while (true) {
-      const currentRecordSchema = messageChain[currentMessageIndex].descriptor.schema!;
-      const currentRecordType = recordSchemaToLabelMap.get(currentRecordSchema)!;
-
-      if (allowedRecordsAtCurrentLevel === undefined || !(currentRecordType in allowedRecordsAtCurrentLevel)) {
-        throw new Error(`record with schema: '${currentRecordSchema}' not allowed in structure level ${currentMessageIndex}`);
-      }
-
-      // if we are looking at the inbound message itself (the last message in the chain),
-      // then we have found the access control object we need to evaluate against
-      if (currentMessageIndex === messageChain.length - 1) {
-        return allowedRecordsAtCurrentLevel[currentRecordType];
+    // get top level rule set
+    const topLevelRuleSet: ProtocolRuleSet | undefined = protocolDefinition.records[protocolPathArray[0]];
+    if (topLevelRuleSet === undefined) {
+      throw new DwnError(DwnErrorCode.ProtocolAuthorizationMissingRuleSet,
+        `No rule set defined for protocolPath ${protocolPathArray[0]}`);
```

Comment:
Question for reviewers:
Previously this error message was:
```
throw new Error(`record with schema: '${currentRecordSchema}' not allowed in structure level ${currentMessageIndex}\`);
```

Do we prefer language similar to that or "No rule set defined" as I have it here?

Reply:
We definitely should avoid reusing error code in principle. I am more interested in knowing why we can't include this check in the loop. (I have not studied the code yet)

Reply:
That seems fine for now, but only because I think I can make it pretty again once I refactor `records` into an array of `ProtocolRuleSet`s.

My gripe is that I don't like having a variable called `currentRuleSet` where the type is not just `ProtocolRuleSet`. And having two different error codes for what is IMO the same error also feels wrong.

Reply:
> the desire to figure out the exact line that causes an issue when debugging

great point, I didn't think of that. You've got me convinced :)